### PR TITLE
New component OSS::NavTab

### PR DESCRIPTION
**Pull request by sBenjamin270495**
### What does this PR do?

Add a new component OSS::NavTab used to switch between many pages / interfaces.

Related to: [#upfluence/oss-component](https://github.com/upfluence/backlog/issues/1725)

### Good PR checklist

- [x] Title makes sense
- [x] Is against the correct branch
- [x] Only addresses one issue
- [x] Properly assigned
- [x] Added/updated tests
- [x] Added/updated documentation
- [ ] Migrated touched components to Glimmer Components
- [x] Properly labeled

### CODE REVIEW

File: addon/components/o-s-s/nav-tab.hbs [hunk 1]
```diff
@@ -0,0 +1,23 @@
+<div class="fx-row fx-1 tab-container">
+  {{#each @tabArray as |tab|}}
+    <div class="fx-col tab {{if tab.selected "tab--selected"}} {{if tab.disabled "tab--disabled"}}"
+         {{on "click" this.onSelectTab}} type="button" tabindex="0">
```
Comment (phndiaye) [inline addon/components/o-s-s/nav-tab.hbs:4]: i think you have to pass the clicked tab as an argument here using: `{{on "click" (fn this.onSelectTab tab)}}`

The `fn` helper creates as function that wraps another one (the first parameter) and pass it the remaining parameters. This way you can have:
```
onSelectTab(selectedTab) {

}
```

File: addon/components/o-s-s/nav-tab.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+export default class OSSNavTab extends Component<OSSNavTabArgs> {
+  constructor(owner: unknown, args: OSSNavTabArgs) {
+    super(owner, args);
+
+    assert(
+      '[component][OSS::NavTab] The parameter @onSelection of type function is mandatory',
+      typeof args.onSelection === 'function'
+    );
+
+    assert('[component][OSS::NavTab] The parameter @tabArray is mandatory', args.tabArray);
```
Comment (phndiaye) [inline addon/components/o-s-s/nav-tab.ts:28]: maybe assert more than `typeof args.tabArray === 'array'` and that the length is `> 0` ?

File: addon/components/o-s-s/nav-tab.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+    let tabArrayValide = true;
+
+    args.tabArray.forEach((element: TabDefinition) => {
+      if (!element.label && !element.icon) {
+        tabArrayValide = false;
+      }
+    });
```
Comment (phndiaye) [inline addon/components/o-s-s/nav-tab.ts:36]: First, you can move this to a `_checkTabArrayValidity` private method
Second, this can be factorise into one line like:
```
!args.tabArray.some((element: TabDefinition) => !element.label && !element.icon)
```

The `Array#some` method returns true if one of the element respects the condition

File: addon/components/o-s-s/nav-tab.hbs [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+      <div class="fx-row tab-content fx-gap-px-9">
+        {{#if tab.icon}}
+          <i class={{tab.icon}} />
+        {{/if}}
+        {{#if tab.label}}
+          <span>{{tab.label}}</span>
+        {{/if}}
+        {{#if tab.infoCircle}}
+          <i class="far fa-info-circle infoCircle" />
```
Comment (phndiaye) [inline addon/components/o-s-s/nav-tab.hbs:13]: `far fa-info-circle` is already a precise enough class. we can use it instead of adding an additiona `infoCircle`

File: addon/components/o-s-s/nav-tab.hbs [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+        {{/if}}
+        {{#if tab.notificationDot}}
+          <div class="iconNotif" />
```
Comment (phndiaye) [inline addon/components/o-s-s/nav-tab.hbs:16]: there is a font awesome icon that is exactly a dot. that one + the `font-color-error-500` class should give you the red dot you want, instead of doing it in css

File: tests/dummy/app/controllers/application.js [hunk 6]
```diff
@@ -138,6 +150,11 @@ export default class ApplicationController extends Controller {
   countDownAction() {
     console.log('countDownAction');
   }
+
+  @action
+  onSelectionNavTab() {
+    console.log('onSelectionNavTab');
```
Comment (Miexil) [inline tests/dummy/app/controllers/application.js:156]: You should test that the tab selection works here, the navtab component should return the selected tab, and you can set it to selected=true from here. (and deselect the previously selected one)

File: addon/components/o-s-s/nav-tab.hbs (hunk 1 above)
Comment (Miexil) [inline addon/components/o-s-s/nav-tab.hbs:4]: ```as |tab index|```

and ```tabindex={{index}}```

File: addon/components/o-s-s/nav-tab.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+    assert(
+      '[component][OSS::NavTab] The @label or @icon parameter is mandatory for each element in @tabArray',
+      tabArrayValide
+    );
+  }
+
+  @action
+  onSelectTab(): void {
+    this.args.onSelection();
```
[the remark this answers is not in the document]
Comment (sBenjamin270495) [reply addon/components/o-s-s/nav-tab.ts:46]: I added a "pointer-event: none" in CSS which cancel the trigger of the function when you click on tab disabled so i don't need to make a condition in the function

[the remark this answers is not in the document]
Comment (sBenjamin270495) [reply addon/components/o-s-s/nav-tab.ts:46]: But i had "disabled={{tab.disabled}}" on button because the keyboard triggered the function, now we can't use keyboard for triggered function

File: addon/components/o-s-s/nav-tab.stories.js [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+  argTypes: {
+    onSelection: {
+      type: { required: true },
+      description: 'A callback triggered when the tab has been clicked.'
+    },
+
+    tabArray: {
+      type: { required: true },
+      description:
+        'Array of TabDefinition which have a parameter: <br/> -icon?: string; <br/> -label?: string; <br/> -infoCircle?: boolean; <br/> -notificationDot?: boolean; <br/> -selected: boolean; <br/> -disabled: boolean; <br/> @label or @icon is mandatory for each element of tabArray'
```
Comment (Miexil) [inline addon/components/o-s-s/nav-tab.stories.js:16]: Array of TabDefinition which has the following parameters:

File: addon/components/o-s-s/nav-tab.stories.js [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    { label: 'Tab', icon: 'far fa-thumbs-up', infoCircle: true, notificationDot: true, selected: true },
+    { label: 'Tab', icon: 'far fa-thumbs-up', infoCircle: true, notificationDot: true, disabled: true },
+    { label: 'Tab', icon: 'far fa-thumbs-up', infoCircle: true, notificationDot: true, selected: true, disabled: true }
+  ],
+  onSelection: action('onSelection')
+};
+
+const Template = (args) => ({
+  template: hbs`
+  <div class="background-color-white padding-px-6"><OSS::NavTab @onSelection={{this.onSelection}} @tabArray={{this.tabArray}} /></div>
```
Comment (Miexil) [inline addon/components/o-s-s/nav-tab.stories.js:38]: could be on multi-line:
```
<div class="background-color-white padding-px-6">
   <OSS::NavTab @onSelection={{this.onSelection}} @tabArray={{this.tabArray}} />
</div>
```

File: app/styles/nav-tab.less [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+    .border-display {
+      align-self: auto;
+    }
+
+    .tab-content {
+      padding: var(--spacing-px-6) var(--spacing-px-18) var(--spacing-px-6) var(--spacing-px-18);
+      height: 34px;
+
+      span {
+        font-size: 14px;
```
Comment (Miexil) [inline app/styles/nav-tab.less:24]: use a preset font class here

File: tests/integration/components/o-s-s/nav-tab-test.ts [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+    });
+
+    test('It throws an error if @tabArray is not passed', async function (assert) {
+      setupOnerror((err: any) => {
+        assert.equal(err.message, 'Assertion Failed: [component][OSS::NavTab] The parameter @tabArray is mandatory');
+      });
+      await render(hbs`<OSS::NavTab @onSelection={{this.onSelection}} />`);
+    });
+
+    test('It throws an error if @label and @icon is missing for any elment in @tabArray', async function (assert) {
```
Comment (Miexil) [inline tests/integration/components/o-s-s/nav-tab-test.ts:43]: `for any element in`

File: tests/integration/components/o-s-s/nav-tab-test.ts [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+      setupOnerror((err: any) => {
+        assert.equal(
+          err.message,
+          'Assertion Failed: [component][OSS::NavTab] The @label or @icon parameter is mandatory for each element in @tabArray'
+        );
+      });
+      await render(hbs`<OSS::NavTab @onSelection={{this.onSelection}} @tabArray={{this.tabArray}} />`);
+    });
+
+    test('Add icon in Tab', async function (assert) {
```
Comment (Miexil) [inline tests/integration/components/o-s-s/nav-tab-test.ts:54]: `Tab Icon displays properly`

File: tests/integration/components/o-s-s/nav-tab-test.ts [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+      assert.dom('.tab-container').exists();
+      assert.dom('.tab-container .tab').exists();
+      assert.dom('.tab-content span').exists();
+      assert.dom('.tab-content i').exists();
+
+      assert.dom('.tab-content i').hasClass('fa-thumbs-up');
+      assert.dom('.tab-content span').hasText('Tab');
+    });
+
+    test('Add info-cirlce in Tab', async function (assert) {
```
Comment (Miexil) [inline tests/integration/components/o-s-s/nav-tab-test.ts:67]: `Tab info-circle displays properly`

File: tests/integration/components/o-s-s/nav-tab-test.ts [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+      assert.dom('.tab-container').exists();
+      assert.dom('.tab-container .tab').exists();
+      assert.dom('.tab-content span').exists();
+      assert.dom('.tab-content i').exists();
+
+      assert.dom('.tab-content i').hasClass('fa-info-circle');
+      assert.dom('.tab-content span').hasText('Tab');
+    });
+
+    test('Add notification dot in Tab', async function (assert) {
```
Comment (Miexil) [inline tests/integration/components/o-s-s/nav-tab-test.ts:80]: `Tab notification dot displays properly`

File: tests/integration/components/o-s-s/nav-tab-test.ts [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+      this.tabArray.push({ label: 'Tab', notificationDot: true });
+      await render(hbs`<OSS::NavTab @tabArray={{this.tabArray}} @onSelection={{this.onSelection}} />`);
+
+      assert.dom('.tab-container').exists();
+      assert.dom('.tab-container .tab').exists();
+      assert.dom('.tab-content .fas.fa-circle').exists();
+    });
+
+    test('Add selected state on Tab', async function (assert) {
```
Comment (Miexil) [inline tests/integration/components/o-s-s/nav-tab-test.ts:89]: `Tab displays selected state properly`

File: tests/integration/components/o-s-s/nav-tab-test.ts [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+      assert.dom('.tab-container').exists();
+      assert.dom('.tab-container .tab').exists();
+      assert.dom('.tab-content span').exists();
+
+      assert.dom('.tab-container .tab').hasClass('tab--selected');
+      assert.dom('.tab-container .tab .border-display').hasStyle({ borderBottom: '2px solid rgb(13, 13, 230)' });
+      assert.dom('.tab-content span').hasText('Tab');
+    });
+
+    test('Add disabled state on Tab', async function (assert) {
```
Comment (Miexil) [inline tests/integration/components/o-s-s/nav-tab-test.ts:102]: `Tab displays disabled state properly`

Comment (Miexil) [reply tests/integration/components/o-s-s/nav-tab-test.ts:102]: Also add a test that checks that the click event is not triggered on click.

- click on disabled tab
- assert that @onSelection has not been called

File: tests/integration/components/o-s-s/nav-tab-test.ts [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+      assert.dom('.tab-container').exists();
+      assert.dom('.tab-container .tab').exists();
+      assert.dom('.tab-content span').exists();
+
+      assert.dom('.tab-container .tab').hasClass('tab--disabled');
+      assert.dom('.tab-content span').hasText('Tab');
+    });
+  });
+
+  test('When clicking on the close icon, the component triggers the @onSelection method', async function (assert) {
```
Comment (Miexil) [inline tests/integration/components/o-s-s/nav-tab-test.ts:115]: Wrong label

File: tests/integration/components/o-s-s/nav-tab-test.ts [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+      });
+      await render(hbs`<OSS::NavTab @tabArray={{this.tabArray}} />`);
+    });
+
+    test('It throws an error if @tabArray is not passed', async function (assert) {
+      setupOnerror((err: any) => {
+        assert.equal(err.message, 'Assertion Failed: [component][OSS::NavTab] The parameter @tabArray is mandatory');
+      });
+      await render(hbs`<OSS::NavTab @onSelection={{this.onSelection}} />`);
+    });
```
Comment (phndiaye) [inline tests/integration/components/o-s-s/nav-tab-test.ts:41]: maybe add a test for @tabArray passed but empty too ?

File: addon/components/o-s-s/nav-tab.stories.js [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+  tabArray: [
+    { label: 'Tab', icon: 'far fa-thumbs-up' },
+    { label: 'Tab2', icon: 'far fa-thumbs-up' },
+    { label: 'Tab3' },
+    { icon: 'far fa-thumbs-up' },
+    { label: 'Tab', icon: 'far fa-thumbs-up', infoCircle: true },
+    { label: 'Tab', icon: 'far fa-thumbs-up', infoCircle: true, notificationDot: true },
+    { label: 'Tab', icon: 'far fa-thumbs-up', infoCircle: true, notificationDot: true, selected: true },
+    { label: 'Tab', icon: 'far fa-thumbs-up', infoCircle: true, notificationDot: true, disabled: true },
+    { label: 'Tab', icon: 'far fa-thumbs-up', infoCircle: true, notificationDot: true, selected: true, disabled: true }
```
[the remark this answers is not in the document]
Comment (sBenjamin270495) [reply addon/components/o-s-s/nav-tab.stories.js:31]: Yes but I put this in tab because we can see that an element selected true and disable true take the disable style.
And this case shouldn't exist too with selected true and disable true.
So are you sure that i delete it ?

[the remark this answers is not in the document]
Comment (phndiaye) [reply addon/components/o-s-s/nav-tab.stories.js:31]: Indeed ! thanks for the explanation :)

**Review by Miexil — APPROVED** [review]

**Review by phndiaye — APPROVED** [review]
